LeanSAT/Data/Cnf.lean

```diff
@@ -0,0 +1,643 @@
+/-
+Copyright (c) 2023 Wojciech Nawrocki. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wojciech Nawrocki, Jeremy Avigad
+-/
+
+import LeanSAT.Upstream.ToMathlib
+import LeanSAT.Model.PropFun
+import LeanSAT.Model.PropVars
+
+namespace LeanSAT
+
+open Model
+
+/-! ### Literals -/
+
+/-- The type `L` is a representation of literals over variables of type `ν`. -/
+@[specialize]
+class LitVar (L : Type u) (ν : outParam $ Type v) where
+  negate : L → L
+  mkPos : ν → L
+  mkNeg : ν → L := fun x => negate (mkPos x)
+  toVar : L → ν
+  polarity : L → Bool
+  -- TODO: ν moreover needs to biject with PNat (perhaps in a separate typeclass)
+  -- so that we can use variable names as array indices
+
+namespace LitVar
+
+def mkLit (L : Type u) {ν : Type v} [LitVar L ν] (x : ν) (p : Bool) : L :=
+  if p then mkPos x else mkNeg x
+
+variable {L : Type u} {ν : Type v} [LitVar L ν]
+
+-- controversial maybe?
+instance : Coe ν L :=
+  ⟨mkPos⟩
+
+instance : Neg L :=
+  ⟨negate⟩
+
+instance [ToString ν] : ToString L where
+  toString l := if polarity l then s!"{toVar l}" else s!"-{toVar l}"
+
+def toPropForm (l : L) : PropForm ν :=
+  if polarity l then .var (toVar l) else .neg (.var $ toVar l)
+
+def toPropFun (l : L) : PropFun ν :=
+  if polarity l then .var (toVar l) else (.var $ toVar l)ᶜ
+
+open PropFun
+
+theorem satisfies_iff {τ : PropAssignment ν} {l : L} :
+    τ ⊨ toPropFun l ↔ τ (toVar l) = polarity l := by
+  dsimp [toPropFun, polarity]
+  aesop
+
+end LitVar
+
+/-! ### Lawful literals -/
+
+-- TODO: see if a shorter list of axioms implies this one
+open LitVar in
+class LawfulLitVar (L : Type u) (ν : Type v) [LitVar L ν] where
+  toVar_negate (l : L) : toVar (negate l) = toVar l
+  toVar_mkPos (x : ν) : toVar (mkPos (L := L) x) = x
+  toVar_mkNeg (x : ν) : toVar (mkNeg (L := L) x) = x
+  polarity_negate (l : L) : polarity (negate l) = !polarity l
+  polarity_mkPos (x : ν) : polarity (mkPos (L := L) x) = true
+  polarity_mkNeg (x : ν) : polarity (mkNeg (L := L) x) = false
+  protected ext (l₁ l₂ : L) : toVar l₁ = toVar l₂ → polarity l₁ = polarity l₂ → l₁ = l₂
+
+namespace LitVar
+variable {L : Type u} {ν : Type v} [LitVar L ν] [LawfulLitVar L ν]
+open LawfulLitVar
+
+attribute [simp] toVar_negate toVar_mkPos toVar_mkNeg polarity_negate polarity_mkPos polarity_mkNeg
+attribute [ext] LawfulLitVar.ext
+
+@[simp] theorem var_neg (l : L) : toVar (-l) = toVar l :=
+  toVar_negate l
+@[simp] theorem polarity_neg (l : L) : polarity (-l) = !polarity l :=
```

Comment:
Ok I tried it, but I think it's in the wrong direction. `polarity_negate` needs to see a `negate` to apply so we'd need `-l = negate l` as a `simp` lemma. Since we don't want that, we need to keep `polarity_neg`.

Reply:
I meant the other way around -- remove `simp` from `polarity_negate`/`toVar_negate`, and add the simp lemma for `negate l = -l` to prefer the notation form. Or maybe it's even better to just use the `Neg` syntax in the def of `LawfulLitVar` so we don't duplicate the lemmas. We can change it later.

Reply:
Ah I see. Just pushed this version to `dev`.